### PR TITLE
Add option for thinLTO

### LINKED ISSUE #7493: Add support for ThinLTO when using Clang and LLD
**Describe the bug**
ThinLTO has significant improvements under the "Full" LTO path.
It scales incredibly well from a memory consumption point of view and gives incremental and threaded linking to a project.
ThinLTO has long been stable with other build systems also supporting it now.
https://clang.llvm.org/docs/ThinLTO.html

**To Reproduce**
Easy enough to reproduce,
`CC=clang CC_LD=lld meson -Db_lto=true <remaining options>`

Meson will only set `-flto` for the compile args in this instance

**Expected behavior**
I expect meson to check which version of clang and lld is being used, and if it is above a certain version (ThinLTO was supported back with clang 3.9, but probably shouldn't be enabled there) then by default enable the thin lto option with `-flto=thin`

Additionally a nice feature might be an option to set LTO flavour. This way people can still build with `-flto=full` for their final releases but keep thin around for their development and regression testing.
`-Db_lto_flavor=full, -Db_lto_flavor=thin`

**system parameters**
* This isn't a cross-building problem
* Linux at the very least effected, MacOS might matter as well, unsure if thinlto is supported under Windows
* Python version doesn't matter
* `git HEAD` for meson, checked the source tree and it isn't supporting thinlto
* Ninja doesn't matter here.

### LINKED ISSUE #7820: meson does not support parallel linking for lto enabled builds
As per GCC documentation, it supports -flto[=n], and If you specify the optional n, the optimization and code generation done at link time is executed in parallel using n parallel jobs by utilizing an installed make program. The environment variable MAKE may be used to override the program used. 

As per meson built-in options b_lto can either be true or a false, and it would be great if meson supports passing "n" as well.

### PR DESCRIPTION
**Pull request by dcbaker**
This adds two separate builtin options. I'm not sure how else to solve this, given that we have two separate toggles to add, and we can't break the existing options.

Fixes #7493
Fixes #7820

### CODE REVIEW

**Comment by jpakkane** [comment]
Could we make it so that we can remove the `b_lto` option later? Such as having also the value `none` for `b_lto_mode` and print a warning whenever someone does `get_option('b_lto')`? Then we could remove it after some releases.

Similarly, could we always determine the number of LTO threads automatically given that we know the maximum number of parallel link jobs, so the number of threads would be num_cores / num_link_jobs? It would work locally at least, not sure about distributed builds, though.

**Comment by dcbaker** [comment]
So my approach with the threads is to use the the number of virtual threads (cores + hyper threads) if the linker doesn't implement it's own auto jobs (basically just old versions of gcc don't). if it does implement it's own, I just let it make it's own decisions, its probably smarter than us anyway. I'm suspect that's going to be more efficient than us trying to guess how many compile/link jobs are going to be running at a time and trying to optimize for that.

I actually like having the mode separate from turning lto on and off, I could do something like this in my clang native file:
```ini
[binaries]
c = ['clang']
cpp = ['clang++']

[built-in options]
b_lto_mode = 'thin'
```
That way I get thin lto whenever lto gets turned on with clang, but i'm not actually turning on lto itself.

**Comment by jpakkane** [comment]
> basically just old versions of gcc don't

In that case could we not remove the option and do the automatic thing and say that it is suboptimal for old versions of GCC. I suspect not that many people are using it anyway.

**Comment by satmandu** [comment]
Could we have a b_lto_mode thin fail over to auto if using gcc? 

In [Chromebrew](https://github.com/skycocker/chromebrew) we are trying to set default options and it would be nice to be able to just set sane optimized defaults for most builds which don't need to be flipped over if we have to use clang.

(ChromeOS is build with libraries which require lld, due to Google's use of ```section `.relr.dyn'``` so  we do have to flip over to using lld a fair bit, and since using LTO with LLD requires clang, it would be nice to just flip over CC/CXX without also having to muck with b_LTO as well for every single package which requires this change.)

**Comment by bonzini** [comment]
Do you get up to M*N processes if you have `ninja -jM` and N LTO jobs?

**Comment by dcbaker** [comment]
@bonzini my understanding is that gcc and clang use threads not processes internally, so you'd end up with M processes and N threads

**Comment by bonzini** [comment]
Should you add a separate pool for linking jobs to avoid running them in parallel?

**Comment by jpakkane** [comment]
There already is one. :D (On the Ninja backend at least.)

**Comment by jpakkane** [comment]
I could not come up with a backwards compatible way of doing this, so let's go with this, with the following:

- if `thin` is not available then fall back to full LTO
- take out the number of lto threads option

The latter can be sort-of-emulated with max linker jobs. If we need something fancier then we probably need to tie these two together somehow (and also with VS and possibly also the VS backend). But we should only do that after we have somewhat wider usage experience and know where the actual bottlenecks are.

Does this sound ok, or are there issues still?

**Comment by nickbroon** [comment]
Can you give an example of controlling the max number of linker threads?

I have a project that links just a single very large executable, using LTO, and the difference between `-flto` and `-flto=[n]` is very noticeable (when setting n to `nproc`).

**Comment by jpakkane** [comment]
From what I can tell the linkers should enable multithreading automatically. If not, then we should set that by default so it works out of the box. So maybe to something like `number_of_processors/max_link_jobs`?

**Comment by dcbaker** [comment]
No, they only enable it if you tell them to AFAICT

**Comment by dcbaker** [comment]
the commit you asked me to drop added support for controlling that, and would give them their `auto` option if you set the number of jobs to `0`, otherwise the number you asked for.

**Comment by jpakkane** [comment]
I misread that then. Well at the very least the behaviour now should be what `auto` should have done.

Related, it would be nice if there was a way to tell Ninja that some process is going to use more cores than one so it could use that info to drive process spawning.

**Comment by dcbaker** [comment]
Okay, threads support added back.

To be clear, the default for threads is now 0, for clang 0 means `pick a number internally`, for gcc we convert that to `auto` if the version of gcc supports that, or `number of logical cores` if not.

**Comment by nickbroon** [comment]
That behaviour for `b_lto_threads` seems reasonable, predictable and without surprise. 👍

**Comment by nickbroon** [comment]
Is there anything else needed for the PR before it can be considered for merge, for 0.57? Are the 2 failed checks expected?

**Comment by jpakkane** [comment]
⚔️ conflicts.

**Comment by dcbaker** [comment]
Fixed.

File: docs/markdown/snippets/lto_threads.md [hunk 1]
```diff
@@ -0,0 +1 @@
+## Supported
```
Comment (jpakkane) [inline docs/markdown/snippets/lto_threads.md:1]: This should probably say something more. :)

**Pull request merged**